tests/playbooks/tests_states.yml

```diff
@@ -4,74 +4,134 @@
+              - tasks/assert_device_absent.yml
+              - tasks/assert_profile_present.yml
+            lsr_cleanup:
+              - tasks/cleanup_profile+device.yml
+      tags:
+        - tests::states:create_without_autoconnect
 
-    # test case (downing a non existent profile)
-    # I can set down and remove a non existent profile
-    # This will be rescued until issue 188 is fixed
-    - name: Verify that the play fails with the expected error message
-      block:
-        - name: Set down non existent profile
-          include_role:
-            name: linux-system-roles.network
+    - block:
+        - include_tasks: tasks/run_test.yml
           vars:
-            network_connections:
-              - name: non_existent
-                state: down
-                persistent_state: absent
+            lsr_description: I can activate an existing profile
+            lsr_setup:
+              - tasks/create_bridge_profile.yml
+            lsr_test:
+              - tasks/activate_profile.yml
+            lsr_assert:
+              - tasks/assert_device_present.yml
+              - tasks/assert_profile_present.yml
+            lsr_cleanup:
+              - tasks/cleanup_profile+device.yml
+      tags:
+        - tests::states:activate
 
-        - name: 'UNREACH'
-          set_fact:
-            __network_test_failed: false
+    - block:
+        - include_tasks: tasks/run_test.yml
+          vars:
+            lsr_description: I can remove an existing profile without taking it
+              down
+            lsr_setup:
+              - tasks/create_bridge_profile.yml
+              - tasks/activate_profile.yml
+            lsr_test:
+              - tasks/remove_profile.yml
+            lsr_assert:
+              - tasks/assert_device_present.yml
+              - tasks/assert_profile_absent.yml
+            lsr_cleanup:
+              - tasks/cleanup_profile+device.yml
+      tags:
+        - tests::states:remove_up
 
-      rescue:
-        - name: Role has failed when it should have
-          set_fact:
-            __network_test_failed: true
+    - block:
+        - include_tasks: tasks/run_test.yml
+          vars:
+            lsr_description: I can take a profile down that is absent
+            lsr_setup:
+              - tasks/create_bridge_profile.yml
+              - tasks/activate_profile.yml
+              - tasks/remove_profile.yml
+            lsr_test:
+              - tasks/remove+down_profile.yml
+            lsr_assert:
+              - tasks/assert_profile_absent.yml
+            lsr_asssert_when:
+              - what: tasks/assert_device_absent.yml
+                when: "{{ network_provider == 'nm' }}"
+            lsr_cleanup:
+              - tasks/cleanup_profile+device.yml
+      tags:
+        - tests::states:remove_down
 
-      always:
-        - assert:
-            that: __network_test_failed
-            fail_msg: "The role did not fail when it should have"
+    - block:
```

Comment:
This whole blocks seems needlessly repetitive. As the profile was already removed and downed in the previous block, can't you simply omit all `lsr_setup:` (the system is already set up properly) and keep only `lsr_test:` and `lsr_assert:`?

Reply:
The repetition is there to make the tests independent from each other to allow to select only one of them with "--tags tests::states:remove_down_twice". I guess a possibility might be to actually join test/assert and have only the "twice" test but since the output from one role invocation is very much, it will then create a lot of logs for just one test.

Reply:
I remember that we did not like Molecule exactly because it did not allow this chaining of tests where the state created by one test was used as setup for another test. If this is not the your priority anymore, we could reconsider the decision to avoid Molecule (ok, there are also other problems).
Question is, whether the ability of selecting just one test via tags is worth the complication (and the increased run time due to repetitive role invocations). I am not sure myself here. So far, we have apparently not been missing the ability to select a subset of a test playbook, but maybe your experience is different.

Reply:
When I add or debug a test, I find it quite useful to be able to select that specific test since running all the tests takes up a lot of time. If it takes longer for the tests to run in the CI, it does not matter that much because this can happen without me watching it.

Also making the tests independent allows to run all the tests instead of aborting after the first failure. This is not possible currently but I try to develop the tests with some idea how to map them to pytest. Then the setup/cleanup would be part of a fixture  and the test/assert part of the actual test. Maybe there will be some opportunities to optimize this then but by writing the tests only in Ansible, optimization seems to be quite hard (when prioritizing readability of tests and logs before execution time).

Reply:
> Also making the tests independent allows to run all the tests instead of aborting after the first failure. This is not possible currently

If they are independent, they could be separated easily into separate playbooks and running all of them will then be easy.